### PR TITLE
add shared storage in windows

### PR DESCRIPTION
**Pull request by yajiedesign**
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage

~~~- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented.~~~


~~~- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change~~~

### Changes ###


## Comments ##

### CODE REVIEW

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ class ConnectionWrapper(object):
     NDArray via shared memory."""
 
     def __init__(self, conn):
-        self.conn = conn
+        self._conn = conn
```
Comment (piiswrong) [inline python/mxnet/gluon/data/dataloader.py]: ?

File: src/storage/cpu_shared_storage_manager.h [hunk 2]
```diff
@@ -91,11 +95,18 @@ class CPUSharedStorageManager final : public StorageManager {
  private:
   static constexpr size_t alignment_ = 16;
 
-  std::mutex mutex_;
+  std::recursive_mutex mutex_;
```
Comment (piiswrong) [inline src/storage/cpu_shared_storage_manager.h]: is this slower?

File: src/storage/cpu_shared_storage_manager.h [hunk 3]
```diff
@@ -91,11 +95,18 @@ class CPUSharedStorageManager final : public StorageManager {
...
   std::mt19937 rand_gen_;
-  std::unordered_map<void*, Storage::Handle> pool_;
+  std::unordered_map<void*, Storage::Handle> pool_; 
+#ifdef _WIN32
+  std::unordered_map<void*, Storage::Handle> is_free_;
+  std::unordered_map<void*, HANDLE> map_handle_map_;
```
Comment (piiswrong) [inline src/storage/cpu_shared_storage_manager.h]: why do you need two more maps?

Comment (yajiedesign) [reply src/storage/cpu_shared_storage_manager.h]: only use pool_ has the problem of repeated release.

File: python/mxnet/gluon/data/dataloader.py (hunk 1 above)
Comment (yajiedesign) [reply python/mxnet/gluon/data/dataloader.py]: __getattr__ is dead cycle call. change the name same time.

**Comment by piiswrong** [comment]
Could you rebase?

**Pull request merged**